aci: extract tar in chroot

sgotti opened the pull request:
this introduces tar extraction inside a chroot. This is done spawning a
new process. Instead of creating a new executable (difficult to find,
package etc...) it adds the concept of a multicall process (like busybox).

Now a new multicall command "extracttar" is added.
tar.ExtractTar executes this new command passing a pipe to send the
filesMap and the tar contents to stdin.

The incomplete "insecure link" checks are removed from the the tar
functions.  Additionally these functions are now not exported and the
extraction directory is always the root directory.

yifan-gu commented on pkg/chroottar/chroottar.go:
```diff
@@ -0,0 +1,99 @@
+}
+
+func extractTar() error {
+	runtime.LockOSThread()
```
Why we still need to lock the thread here? Isn't this function running in another process?

yifan-gu replied:
Or are you preventing other goroutines from chrooting?

sgotti replied:
@yifan-gu right. As chroot should be per process, the need to lock a goroutine to a specific thread shouldn't be needed. @vcaputo any thoughts?

sgotti commented:
@jonboulle @vcaputo @yifan-gu Thanks for your review. I tried implementing all your comments. Additionally I removed the runtime.LockOSThread() call. In one of the replies I added some note on handling subprocess hangs/killing/timeouts.

Unfortunately CAP_SYS_CHROOT isn't provided by the CI tools, perhaps the CI user will be able to set that capability on the executable (quite difficult to do with go test as it compile and executes...).
Another possibility will be to fallback to in process tar extraction during tests, but this will not cover the reality... @alban?

jonboulle commented on pkg/chroottar/chroottar.go:
```diff
@@ -0,0 +1,104 @@
+	cmd.Stdin = rs
+	var encodeErr error
+	go func() {
+		if err := enc.Encode(pwl); err != nil {
```
`encodeErr = enc.Encode(pwl)` is simpler and should be the same

sgotti replied:
:exclamation: :disappointed: :sweat_smile:

sgotti commented:
I pushed a new version that handles the pwl encoding return error inside the goroutine with a channel, this will avoid a race (reported by the race detector) that will bring to read the encodeErr variable before the encoding has returned.
Additionally, I added a function to disable doing the chroot in the subprocess. This function is called by the init functions of the store tests. The subprocess spawning and everything else remains the same, just the call to chroot and chdir are skipped in the subprocess (and the dir is set to the extraction dir instead of "/"). This is the best I can find to do the less different code path and make CI happy.

alban commented:
@sgotti : the functional tests in Semaphore are executed with `sudo` so we have `CAP_SYS_CHROOT`:
https://github.com/coreos/rkt/blob/master/tests/test#L21

But the unit tests are executed as the normal user without `CAP_SYS_CHROOT`.

I am not sure what's best. The unit tests in Semaphore could be run as root as well by changing `./test` to `./sudo test` in `tests/run-build.sh`:
https://github.com/coreos/rkt/blob/master/tests/run-build.sh#L20

sgotti commented:
@alban Thanks! that's good. Instead with travis there's probably no way. I'm thinking to disable the chroot ing reading a flag passed to go test, so we can just change `.travis.yml` to call something like `./test --nochroot`.

yifan-gu commented on pkg/chroottar/chroottar.go:
```diff
@@ -0,0 +1,121 @@
+	}
+
+	extractDir := "/"
+	if chroot {
```
How is this compared to hide this with an interface, e.g.

```go
type DirSetter interface {
    SetDir(dir string)
}
```

Which we can do chroot(dir), chdir("/") in real binary, and do chdir(dir) in unit tests.

sgotti replied:
@yifan-gu it will be probably cleaner. My doubt is how to pass that interface to extractTar as its called in the subprocess. Perhaps more than an interface we can declare e DirSetter function pointer that it can be overridden by the tests.
Another problem is that pkg/tar.ExtractTar needs `dir` to be an absolute path (I think that a check should be added or if a relative path is given we can generate an absolute path joining it with the CWD), or the checks for the insecure link will probably fail. So, at the moment looks like its not possible to do a chdir(dir) leaving extractDir="/".

jonboulle commented on pkg/multicall/multicall.go:
```diff
@@ -0,0 +1,97 @@
+// If the functions returns an error, it will be printed to stderr and will
+// exit with an exit status of 1, elsewhere it will return with a 0 exit
+// status.
+func Exec() {
```
`MaybeExec` is probably more accurate?

sgotti replied:
definitely.

jonboulle commented on pkg/multicall/multicall.go:
```diff
@@ -0,0 +1,97 @@
+// Exec should be called at the start of the program, if the process argv[0] is
+// a name registered with multicall, the related function will be executed.
+// If the functions returns an error, it will be printed to stderr and will
+// exit with an exit status of 1, elsewhere it will return with a 0 exit
```
s/elsewhere/otherwise/
Also "exit" rather than "return'

jonboulle commented on pkg/chroottar/chroottar.go:
```diff
@@ -0,0 +1,121 @@
+	ptar "github.com/coreos/rkt/pkg/tar"
+)
+
+var testing bool
```
unused?

jonboulle commented on pkg/multicall/multicall.go:
```diff
@@ -0,0 +1,97 @@
+	}
+}
+
+// GetEntrypoint returns the Entrypoint related to the requested command name.
```
any intention for this to be used?

sgotti replied:
At the moment no. I just put it if in a future there'll be the need to call a command outside the package that declared it.

Would you prefer if I remove it for now (it will be re added when needed)?

jonboulle replied:
Yes please

jonboulle commented on store/store_test.go:
```diff
@@ -27,10 +27,17 @@ import (
 const tstprefix = "store-test"
 
+func init() {
+	chroottar.SetChroot(false)
```
a thought, should we check for CAP_SYS_CHROOT (or root) before doing this? e.g.

sgotti commented:
Updated with your review. Now the build/tests on semaphoreci are executed as root.
Only on travis chroot is disabled. To disable the chroot an environment variable CHROOT=false is defined in `.travis.yml` before executing `./test`.

An user testing locally without the CAP_SYS_CHROOT, at the moment, needs to manually export it.

Would you prefer if the capability (or root) is detected by the test script that will automatically set CHROOT=false (perhaps printing a warning?)

jonboulle commented:
@sgotti I think this is inverted; tests should be runnable as non-root by default.
My thinking was that we could do it in the test itself rather than in the script - i.e. during init(), check for root/capability (`geteuid() == 0 || has_effective_cap(CAP_SYS_CHROOT)`), and then decide on that basis whether to test with chroot or not (actually ideally root would test both cases)

sgotti commented:
@jonboulle yeah. completely agree. After 3 runs of `./test` where i forgot to export the env variable I'm just starting hating myself!

jonboulle commented:
:stuck_out_tongue:

sgotti commented:
@jonboulle Done. It checks the CAP_SYS_CHROOT (as also root should have dropped it) of the current process, if retrieving the capabilities fails for any reason it falls back checking the effective uid.

jonboulle commented on store/store_test.go:
```diff
@@ -26,11 +26,32 @@ import (
+	if err == nil {
+		chroot = caps.Get(capability.EFFECTIVE, capability.CAP_SYS_CHROOT)
+	} else {
+		if os.Geteuid() == 0 {
```
nit, just oneline it - `chroot = os.Geteuid() == 0`

vcaputo commented:
lgtm, thanks sgotti!

jonboulle commented:
After talking about this more with Vito OOB I think we should be even stricter about this, and follow up with:
a) removing the existing ["insecure links"](https://github.com/coreos/rkt/blob/e5be761deb0620403aa181702b05609f994cba9a/pkg/tar/tar.go#L33) code
b) removing the [non-chrooted](https://github.com/coreos/rkt/pull/923/files#diff-c21b6cd6b3053b06d65ab9f73253670bR70) version of the extraction (in this diff)
c) either being very explicit that [ExtractTar](https://github.com/coreos/rkt/blob/e5be761deb0620403aa181702b05609f994cba9a/pkg/tar/tar.go#L44) (and by extension [ExtractFile](https://github.com/coreos/rkt/blob/e5be761deb0620403aa181702b05609f994cba9a/pkg/tar/tar.go#L88)) are not secure and should _only_ be called from within a chroot - or actually just merging them into extracttar and not exposing it elsewhere (for example we could just remove the "dir" argument)
d) skipping the tests entirely when run as non-root

@sgotti does this sound reasonable? and/or would you like to take it on? :-) no pressure.

sgotti commented:
@jonboulle @vcaputo Seems good.

> c) either being very explicit that ExtractTar (and by extension ExtractFile) are not secure and should only be called from within a chroot - or actually just merging them into extracttar and not exposing it elsewhere (for example we could just remove the "dir" argument)

I'm quite for merging it (perhaps keeping only `pkg/tar` instead of `pkg/chroottar`).

Would you prefer to do the changes in this PR (seems better if you're ok in merging inside pkg/tar to avoid an additional file or lines move in another commit)?

jonboulle commented:
@sgotti If you don't mind doing it in this PR, that would be awesome.

sgotti commented:
Updated. PTAL.

jonboulle commented on pkg/tar/tar_test.go:
```diff
@@ -158,67 +178,12 @@ func checkExpectedFiles(dir string, expectedFiles map[string]*fileInfo) error {
-		}
+func TestExtractTarFolders(t *testing.T) {
+	if !chroot {
+		t.Logf("chroot capability not available. Disabling tar tests\n")
```
t.Skipf perhaps?

sgotti replied:
Uh!

jonboulle commented on pkg/tar/chroot.go:
```diff
@@ -0,0 +1,107 @@
+	return nil
+}
+
+// ExtractTar extracts a tarball (from a io.Reader) into the given directory
```
two things:
1) please add a note that this will chroot and hence requires privileges
2) should we add a check here for that priv? maybe that's slightly more elegant than failing during the `Chroot()`? I'm on the fence

jonboulle commented:
couple of comments but looks great

sgotti commented:
@jonboulle Done! To avoid copying the same code three times I added a new function HasChrootCapability in pkg/sys/capability.go (don't know if it's ok or will be better to put it in a util package that atm doesn't exists).

jonboulle commented:
![lgtm](http://vignette3.wikia.nocookie.net/icarly/images/e/e0/0001st50.gif/revision/latest?cb=20110625131858)


@vcaputo mind giving a final review?

yifan-gu commented:
@jonboulle I accidently deleted your gif.... sorry!

jonboulle commented:
:-(

On Mon, May 25, 2015 at 3:36 PM, Yifan Gu <notifications@github.com> wrote:

> @jonboulle <https://github.com/jonboulle> I accidently deleted your
> gif.... sorry!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/coreos/rkt/pull/923#issuecomment-105328744>.
>

alban commented:
> @jonboulle I accidently deleted your gif.... sorry!

Let me copy @jonboulle's comment from the email notification:

---

![lgtm](http://vignette3.wikia.nocookie.net/icarly/images/e/e0/0001st50.gif/revision/latest?cb=20110625131858)


@vcaputo mind giving a final review?

vcaputo commented:
looking...

vcaputo commented on pkg/multicall/multicall.go:
```diff
@@ -0,0 +1,88 @@
+func MaybeExec() {
+	name := os.Args[0]
+	if fn, ok := commands[name]; ok {
+		err := fn()
```
any particular reason this isn't in the form of:
```
if err := fn(); err != nil {
        fmt.Fprintln(os.Stderr, err)
        os.Exit(1)
}
```

sgotti replied:
none. Just forgot to compact them.

vcaputo commented on pkg/tar/chroot.go:
```diff
@@ -0,0 +1,113 @@
+
+func extractTarCommand() error {
+	if len(os.Args) < 3 {
+		return fmt.Errorf("not enough arguments")
```
Is there any reason we should ignore excess arguments here?
We should probably fail hard if arguments != 3, it shouldn't happen.

sgotti replied:
You're right!

vcaputo commented:
small nits, otherwise lgtm.

jonboulle commented on pkg/tar/chroot.go:
```diff
@@ -0,0 +1,113 @@
+
+func extractTarCommand() error {
+	if len(os.Args) != 3 {
+		return fmt.Errorf("not enough arguments")
```
nit, now this should say "incorrect number of arguments". maybe it can also print what it expects in that case..?

sgotti replied:
Whoops! We can also write something like `Usage: extracttar DIR OVERWRITE` but it should never be directly executed.

jonboulle replied:
Understood, but still worth it for it to be selfdocumenting IMHO

sgotti replied:
Right! Is it good enough `Usage: extracttar DIR {true|false}`? As for simplicty I avoided the use of flags I'm not sure which is the best way to explain that the second parameter is for enabling/disabling files overwrite.

jonboulle replied:
That seems fine, thanks

jonboulle commented:
@sgotti sorry, one more thing :P

sgotti commented:
@jonboulle Done!

jonboulle commented:
Awesome, thanks so much @sgotti
